### DISCUSSION TITLE
How to set Authorization headers asynchronously

### DISCUSSION THREAD

**Discussion by MattCheetham** [kean/Nuke#820] [Q&A] [answered]
I'm currently using Kingfisher and trying to switch to Nuke. All is going well except I'm currently stuck on how to correctly add Authorization headers to my requests. In Kingfisher I'm able to set a `requestModifier` on the shared Kingfisher object that allows me to modify any requests before they go out, but I'm struggling to work this out with Nuke.

I came across https://github.com/kean/Nuke/issues/331 which was a step in the right direction except setting the header in this way only sets it on first call and then the Authorization header is fixed. This is a problem as our token expires and has to be refreshed by an async method.

Ideally I need a way of intercepting each request that Nuke is about to make, and calling my async method that handles token retrieval/refresh and then returns that token to Nuke to be set on the `Authorization` header.

I'm sure this must be doable as this is a pretty common requirement for requests.

Any help would be hugely appreciated.

**Comment by obrhoff** [comment]
Would something like this not work (not tested)

```
import Nuke
import NukeUI
import SwiftUI

public struct ImageView<Placeholder: View>: View {
    @State private var request: ImageRequest?
    private let placeholder: () -> Placeholder
    private let url: URL?

    public init(
        url: URL? = nil,
        placeholder: @escaping (() -> Placeholder) = { Image.placeholder.resizable() },
    ) {
        self.url = url
        self.placeholder = placeholder
    }

    public var body: some View {
        LazyImage(request: request) { state in
            if let image = state.image {
                image
                    .resizable()
                    .scaledToFill()
            } else {
                placeholder()
            }
        }
        .task {
            // fetch the token here
            guard let url, let accessToken = try? await authenticationManager.accessToken() else {
                return
            }
            
            var urlRequest = URLRequest(url: url)
            urlRequest.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
            self.request = ImageRequest(urlRequest: urlRequest)
        }
    }
}

```

**Reply by kean** [reply, marked as answer]
There is a new async `func willLoadData` method in`ImagePipeline.Delegate` added in the upcoming major release (Nuke 13.0). More details in https://github.com/kean/Nuke/issues/772#issuecomment-4062962638. I believe it should cover this use-case. Let me know if there is anything more I could add to support it better.
